src/widget.c

```diff
@@ -161,6 +184,37 @@ ZMK_LISTENER(led_battery_listener, led_battery_listener_cb);
 ZMK_SUBSCRIPTION(led_battery_listener, zmk_battery_state_changed);
 #endif // IS_ENABLED(CONFIG_ZMK_BATTERY_REPORTING)
 
+#if SHOW_LAYER_COLORS
+uint8_t led_current_layer_color = 0;
+
+void update_layer_color(void) {
+    uint8_t index = zmk_keymap_highest_layer_active();
+
+    if (led_current_layer_color != layer_color_idx[index]) {
+        led_current_layer_color = layer_color_idx[index];
+        struct blink_item color = {.duration_ms = 5, .color = led_current_layer_color, .sleep_ms = 5};
+        LOG_INF("Setting layer color to %s for layer %d", color_names[led_current_layer_color], index);
+        k_msgq_put(&led_msgq, &color, K_NO_WAIT);
```

Comment:
I think this doesn't do much besides making the thread run again and update with the current color, right?

Reply:
Because I'm not an expert in C or Zephyr, I was afraid to control the LED devices from two different threads, so anything that needs to update the leds, I put into the led_msgq. If we don't have to worry about threads and etc, we could just update the LEDs from here, since we aren't performing a blink. By using the queue, though, we can also interact more cleanly with any concurrent blinks.

Reply:
That's fair, depending on the thread is a good idea, yes. I am just wondering if we should have a "dummy" blink type, where it doesn't do anything but only updates the color. Maybe if `blink_ms` is zero, what do you think?

Reply:
Yes, that's essentially what I did, using the existing mechanisms and setting the duration_ms and sleep_ms to 5ms each, which is as close to 0 as I was comfortable getting without knowing if there were any downsides to rapidly toggling the led.

There is existing code that checks for sleep_ms being omitted, and we could add something similar for duration_ms being omitted, and if so, skip the wait state. That way we wouldn't have to add a new attribute to the struct.

Reply:
I think that was the right call. Once I started to separate the different paths, the logic became much cleaner, and I was able to refactor to make it much more obvious what is going on.